File: JavaScript/Core0/OES-js-Core0/simulator.js
/* Changes from OESjs
- sim.objects is a JS Map
 */

/*******************************************************
 General Variables and Functions
 ********************************************************/
// A map of all simulation objects
sim.objects = new Map();
// The Future Events List
sim.FEL = new EventList();
// The statistics variables map
sim.stat = Object.create(null);

/*******************************************************************
 * Initialize a (standalone or experiment scenario) simulation run *
 *******************************************************************/
sim.initializeScenarioRun = function () {
  sim.step = 0;  // simulation loop step counter
  sim.time = 0;  // simulation time
  // set default endTime
  sim.endTime = sim.scenario?.simEndTime || sim.endTime || Number.MAX_SAFE_INTEGER;
  // get ID counter from simulation scenario, or set to default value
  sim.idCounter = sim.scenario?.idCounter || 1000;
  // set up initial state
  if (sim.scenario.setupInitialState) sim.scenario.setupInitialState();
  if (sim.model.setupStatistics) sim.model.setupStatistics();
};
/*******************************************************
 Advance Simulation Time
 ********************************************************/
sim.advanceSimulationTime = function () {
  sim.nextEvtTime = sim.FEL.getNextOccurrenceTime();  // 0 if there is no next event
  // increment the step counter
  sim.step += 1;
  // advance simulation time
  if (sim.nextEvtTime > 0) sim.time = sim.nextEvtTime;
}
/*******************************************************
 Run a simulation scenario
 ********************************************************/
sim.runScenario = function () {
  // Simulation Loop
  while (sim.time < sim.endTime && !sim.FEL.isEmpty()) {
    // if not executed in a JS worker, create simulation log
    if (typeof WorkerGlobalScope === 'undefined' && simLogTableEl) logSimulationStep( simLogTableEl);
    sim.advanceSimulationTime();
    // extract and process next events
    let nextEvents = sim.FEL.removeNextEvents();
    for (let e of nextEvents) {
      // apply event rule
      let followUpEvents = e.onEvent();
      // schedule follow-up events
      for (let f of followUpEvents) {
        sim.FEL.add( f);
      }
      let EventClass = e.constructor;
      // test if e is an exogenous event
      if (EventClass.recurrence) {
        // create and schedule next exogenous events
        sim.FEL.add( e.createNextEvent());
      }
    }
  }
  if (sim.model.computeFinalStatistics) sim.model.computeFinalStatistics();
}
/*******************************************************
 Run an Experiment (in a JS worker)
 ********************************************************/
sim.runExperiment = function (exp) {
  // initialize replication statistics record
  if (sim.model.setupStatistics) sim.model.setupStatistics();
  exp.replicStat = Object.create(null);  // empty map
  for (let varName of Object.keys( sim.stat)) {
    exp.replicStat[varName] = [];  // an array per statistics variable
  }
  // run experiment scenario replications
  for (let k=0; k < exp.nmrOfReplications; k++) {
    sim.initializeScenarioRun();
    sim.runScenario();
    // store replication statistics
    Object.keys( exp.replicStat).forEach( function (varName) {
      exp.replicStat[varName][k] = sim.stat[varName];
    });
  }
  // define exp.summaryStat to be a map for the summary statistics
  exp.summaryStat = Object.create(null);
  // aggregate replication statistics in exp.summaryStat
  Object.keys( exp.replicStat).forEach( function (varName) {
    exp.summaryStat[varName] = Object.create(null);  // empty map
    Object.keys( math.stat.summary).forEach( function (aggr) {
      var aggrF = math.stat.summary[aggr].f;
      exp.summaryStat[varName][aggr] = aggrF( exp.replicStat[varName]);
    });
  });
  // send experiment statistics to main thread
  self.postMessage({experiment: exp});
};

File: Python/Core0/OESpy-Core0/simulator.py
import sys, os
from simulatorUI import logSimulationStep
from oes_foundation import defaults

module_path = os.path.abspath('lib/')
sys.path.insert(1, module_path)
from EventList import EventList

def has_method( o, name):
    return callable( getattr( o, name, None))

class Simulator:
    def __init__(self):
        self.step = 0
        self.time = 0
        self.endTime = 0
        self.nextMomentDeltaT = None
        self.model = None
        self.scenario = Scenario()
        self.config = Configuration( createLog = False)
        self.FEL = EventList()
        self.objects = dict()
        self.stat = None
        self.nextEvtTime = 0
        self._idCounter = 0
    pass

    @property  # define getter for property
    def idCounter(self):
        self._idCounter += 1  # increment on read
        return self._idCounter
    @idCounter.setter
    def idCounter(self, x):
        self._idCounter = x

    def initializeSimulator(self):
        if self.model.time == "discrete": self.nextMomentDeltaT = 1
        else: self.nextMomentDeltaT = self.model.nextMomentDeltaT or defaults["nextMomentDeltaT"]
    
    def initializeScenarioRun(self):
        self.objects.clear()
        self.step = 0
        self.time = 0
        self.endTime = self.scenario.durationInSimTime or float('inf')  # infinity
        self._idCounter = self.scenario.idCounter or 1000
             
        self.scenario.setupInitialState()
        
        if has_method( self.model, "setupStatistics"): self.model.setupStatistics()
    
    def advanceSimulationTime(self):
        nextEvtTime = self.FEL.getNextOccurrenceTime()  # 0 if there is no next event
        self.step += 1
        if nextEvtTime > 0: self.time = nextEvtTime
    
    def runScenario(self):
        while self.time < self.endTime and not self.FEL.isEmpty():
            if self.config.createLog: logSimulationStep(self)
            self.advanceSimulationTime()
            nextEvents = self.FEL.removeNextEvents()
            for e in nextEvents:
                followUpEvents = e.onEvent( self)
                for f in followUpEvents: self.FEL.add(f)
                # check if e is a recurrent/exogenous event
                if has_method( e, "recurrence"):
                    if has_method( e, "createNextEvent"): self.FEL.add( e.createNextEvent( self))
                    else: self.FEL.add( type(e)(delay=e.recurrence()))
        
        if has_method( self.model, "computeFinalStatistics"):
            self.model.computeFinalStatistics()
            for k,v in self.stat.items():
                if type(v) == float: self.stat[k] = round( v, defaults['expostStatDecimalPlaces'])
            print ("\n-------------------- Final Statistics --------------------------")
            print ("Stat:", self.stat)
            
    def runStandaloneScenario(self):
        self.initializeSimulator()
        self.initializeScenarioRun()
        self.runScenario()
    
class Model:
    def __init__(self, time, timeUnit, objectTypes, eventTypes, nextMomentDeltaT=None):
        self.time = time
        self.timeUnit = timeUnit
        self.objectTypes = objectTypes
        self.eventTypes = eventTypes
        self.nextMomentDeltaT = nextMomentDeltaT

class Scenario:
    def __init__(self):
        self.durationInSimTime = 0
        self.idCounter = None
    
    def setupInitialState(self):
        return False

class Configuration:
    def __init__(self, createLog):
        self.createLog = createLog
